Save models before exiting after receiving signal SIGTERM

ugermann opened the pull request:
Some of us run Marian on a cluster with time limits for jobs. When a training job gets killed due to timeout, we lose all the progress since the last checkpoint save (--save-freq=...). This branch / pull request installs a signal handler in class Scheduler that triggers a model save before exit after receiving a SIGTERM (i.e., the default signal for kill <pid>), so that we can pick up where we left it after a training process was killed.

I've tested it to some degree but not exhaustively. Feedback welcome.

snukky commented:
Thanks, I'll take a look. Do regression tests pass with your branch?

ugermann commented:
How do I run them?

On Mon., 29 Apr. 2019, 09:54 Roman Grundkiewicz, <notifications@github.com>
wrote:

> Thanks, I'll take a look. Do regression tests pass with your branch?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/marian-nmt/marian-dev/pull/442#issuecomment-487504697>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABCAWO2DMHJGDP2JTURDQ2LPS2ZV5ANCNFSM4HI5GKVA>
> .
>

snukky commented:
Please follow https://github.com/marian-nmt/marian-regression-tests/blob/master/README.md

ugermann commented:
```
Skipped:
  - tests/training/features/exp-smoothing/test_expsmooth_sync.sh
  - tests/training/restoring/optimizer/test_adam_params_async.sh
  - tests/training/restoring/optimizer/test_adam_params_sync.sh
  - tests/training/restoring/exp-smoothing/test_expsmooth.sh
  - tests/training/restoring/exp-smoothing/test_expsmooth_s2s.sh
  - tests/training/restoring/exp-smoothing/test_expsmooth_sync.sh
  - tests/training/restoring/multi-gpu/test_async.sh
  - tests/training/restoring/multi-gpu/test_sync.sh
  - tests/training/multi-gpu/test_async_sgd_runs.sh
  - tests/training/multi-gpu/test_sync_sgd.sh
  - tests/examples/iris/test_iris.sh
  - tests/examples/mnist/test_mnist_ffnn.sh
  - tests/server/test_ende.sh
  - tests/server/test_ende_align.sh
  - tests/server/test_ende_batch32.sh
  - tests/server/test_ende_cpu.sh
  - tests/server/test_ende_with_empty_lines.sh
---------------------
Ran 157 tests in 00:31:58.166s, 140 passed, 17 skipped, 0 failed
```

emjotde commented:
I assume this leaves SIGINT alone?

emjotde commented:
This has surprisingly many changes that seem unrelated to the SIGTERM issue, no?

snukky commented:
Maybe thre should be an option to turn it off (or on), as sometimes automatic saving might be an unwanted behaviour?

ugermann commented:
@emjotde: In order to understand what was going on, I refactored the code a bit ti make it more readable, apart form mere reformatting of some lambda functions, I moved blocks of code into separate auxiliary functions (e.g., initMPI, prepareData) to declutter the functions that originally contained those code blocks. And yes, this leaves SIGINT alone.

@snukky Making this configurable is trivial, but I'd prefer to have #488 merged first before I make this configurable. In principle, SIGTERM is intended for shutdown with cleanup, and SIGKILL for immediate exit without cleanup.

frankseide commented:
Similar changes may have been made in our internal code. @emjotde, are the changes re variable-sized batches already in public master?

ugermann commented:
Can we pretty-please merge this soon? It's been more than four months since the original filing.

emjotde commented:
I'm merging in our internal branch now. Will take another look at this afterwards. I am a bit wary because of the changes not related to SIGTERM.

emjotde commented:
Looking also at #488

ugermann commented:
I've minimized the differences between master and this branch. Can this be merged, please? This pull request has been waiting for Godot for more than half a year now.

emjotde commented:
Will take a look tomorrow.

emjotde commented on src/training/communicator.h:
```diff
@@ -175,7 +175,17 @@ class DefaultCommunicator : public ICommunicator {
+    };
+
     foreach(scatter);
+    foreach(reset);
```
Where does this change come from?

ugermann replied:
By cherry-picking communicator.h from master in the latest merge with master. The change comes from 
https://github.com/marian-nmt/marian-dev/commit/2111c280362fcf1b76bb13688e5a8f372e435479

ugermann replied:
I actually don't understand why github lists this as a difference.

emjotde replied:
I have been running in to this recently. Seems merge diffs are done against last common ancestor or something. Makes no sense to me either. At least give me a way to see an actual 3-way diff, not a fictional one.

emjotde commented:
Running full regressions tests now, and playing around with it.

emjotde commented on src/training/scheduler.cpp:
```diff
@@ -0,0 +1,32 @@
+
+void
+Scheduler::
+signalHandler_(int sig) {
```
Odd formatting, please keep this in one line.

emjotde commented on src/training/scheduler.cpp:
```diff
@@ -0,0 +1,32 @@
+
+void
+Scheduler::
+installSignalHandlers_() {
```
As above.

emjotde commented on src/training/scheduler.cpp:
```diff
@@ -0,0 +1,32 @@
+  // in signal.h), so we have to do custom log messages here.
+  switch (sig) {
+  case SIGTERM: // save models and exit
+    LOG(info, "[training] Scheduler received signal SIGTERM");
```
indent case

emjotde commented on src/training/scheduler.h:
```diff
@@ -19,6 +19,10 @@ class Scheduler : public TrainingObserver {
 
+  static bool sigterm_;
+  void installSignalHandlers_();
+  static void signalHandler_(int sig);
```
Please add comments to all of these.

emjotde replied:
Also, we do not use _ for function names.

emjotde commented on src/training/scheduler.h:
```diff
@@ -146,9 +150,14 @@ class Scheduler : public TrainingObserver {
       : options_(options), state_(state) {
     ABORT_IF(state_->factor != 1, "state.factor unexpectedly not 1 at this point??");
     updateLearningRate(*state);
+    installSignalHandlers_();
```
This seems like an odd place to put that and rather not well motived. @snukky can you suggest a better location?

snukky replied:
It's in the constructor, so should be OK.

emjotde replied:
Actually, ignore that comment. GitHub cut off the function name so weirdly that I thought it was inside getScheduledLRDecayFactor().

emjotde replied:
The constructor is fine.

emjotde commented on src/training/scheduler.h:
```diff
@@ -175,7 +184,13 @@ class Scheduler : public TrainingObserver {
   void started() { LOG(info, "Training started"); }
-  void finished() { LOG(info, "Training finished"); }
+  void finished() {
+    if (keepGoing(false)) // false means: ignore sigterm flag
```
```keepGoing(/*checkForSigTerm=*/false)```

emjotde commented on src/training/scheduler.h:
```diff
@@ -204,7 +219,8 @@ class Scheduler : public TrainingObserver {
     // Do not validate if already validated (for instance, after the model is
     // loaded) or if validation is scheduled for another update
-    if(state_->validated
+    if(sigterm_
```
Add comment how this is affected by sigterm_

emjotde commented:
General functionality is nice. Have a few requests for changes. Will run regression tests afterwards once more.

emjotde commented:
Regression tests all pass.

snukky commented on src/training/scheduler.cpp:
```diff
@@ -0,0 +1,32 @@
+Scheduler::
+installSignalHandlers_() {
+  // TODO: use sigaction instead of signal
+  signal(SIGTERM, Scheduler::signalHandler_);
```
Why `sigaction` would be better than `signal` and what is the reason it's only a TODO? A quick googling suggests that `sigaction` is not supported on Windows?

ugermann replied:
> Why sigaction would be better than signal
https://stackoverflow.com/questions/231912/what-is-the-difference-between-sigaction-and-signal

> what is the reason it's only a TODO
Convenience of implementation. Old dogs, new tricks and so forth.

snukky replied:
OK, would you mind adding this link into that TODO for futher reference? Thanks

snukky commented on src/training/scheduler.h:
```diff
@@ -175,7 +184,13 @@ class Scheduler : public TrainingObserver {
   void started() { LOG(info, "Training started"); }
-  void finished() { LOG(info, "Training finished"); }
+  void finished() {
+    if (keepGoing(false)) // false means: ignore sigterm flag
```
Is this the only use of `keepGoing(false)`, i.e. only for logging a different message? If so, I would suggest removing the argument from `keepGoing()` (and keeping it simple) and adding here `if(sigterm_)` or similar.

ugermann replied:
It is possible that both are true: SIGTERM recevied and training finished. In that case, we want to log "Training finished.", not "Training interrupted.", so we'd have to say
```
if(sigterm&&!keepGoing()){
...
```
And that's the opposite of "keeping things simple".

snukky replied:
OK, but `keepGoing()` is a part of the public interface of the class, so I would value its simpler version more than a simpler if-statement hidden in a function (plus a comment). It also seems the argument is used only internally inside the class.

emjotde replied:
I think @snukky is right here. If that's only for logging purposes just checking `sigterm_` seems fine.

emjotde commented:
Let us know when you are ready.

ugermann commented:
Compiles now. Didn't run regtests b/c I get different results on different machines.

emjotde commented:
OK, will try again.

emjotde commented:
Not compiling for me, did you push?

emjotde commented:
```
/home/marcinjd/marian-dev/src/training/scheduler.cpp: In member function ‘void marian::Scheduler::installSignalHandlers()’:
/home/marcinjd/marian-dev/src/training/scheduler.cpp:25:19: error: ‘signalHandler_’ is not a member of ‘marian::Scheduler’
   signal(SIGTERM, Scheduler::signalHandler_);
```

ugermann commented:
Weird. Did compile even with make -B. Fixed bug now, methinks. Try again.

emjotde commented:
Looks good, running tests.

emjotde commented:
I am satisfied. Regression tests pass. @snukky, your turn.
